network/omaha/server.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package omaha
+
+import (
+	"encoding/xml"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/coreos/mantle/Godeps/_workspace/src/gopkg.in/tylerb/graceful.v1"
+)
+
+func NewServer(addr string) (*Server, error) {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	srv := &graceful.Server{
+		Server: &http.Server{
+			Addr: addr,
+		},
+	}
+
+	s := &Server{
+		Addr: addr,
+		l:    l,
+		srv:  srv,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/v1/update/", s.UpdateHandler)
+
+	s.srv.Handler = mux
+
+	return s, nil
+}
+
+type Server struct {
+	Addr string
+	l    net.Listener
+	srv  *graceful.Server
+
+	reqChanLock sync.Mutex
+	reqChan     chan *Request
+}
+
+func (s *Server) Serve() error {
+	err := s.srv.Serve(s.l)
+	if err != nil {
+		netErr, ok := err.(*net.OpError)
+		if ok {
+			if netErr.Err.Error() == "use of closed network connection" {
+				return nil
+			}
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+func (s *Server) Stop() {
+	ch := s.srv.StopChan()
+	s.srv.Stop(10 * time.Second)
+	<-ch
+
+	s.reqChanLock.Lock()
+	if s.reqChan == nil {
+		close(s.reqChan)
+	}
+	s.reqChanLock.Unlock()
+}
+
+// RequestChan returns a channel that will receive decoded omaha requests.
+//
+// XXX: this can only be called once per Server instance, otherwise the
+// receivers will miss requests.
+func (s *Server) RequestChan() <-chan *Request {
+	s.reqChanLock.Lock()
+	if s.reqChan == nil {
+		s.reqChan = make(chan *Request)
+	}
+	s.reqChanLock.Unlock()
+	return s.reqChan
+}
+
+func (s *Server) UpdateHandler(rw http.ResponseWriter, r *http.Request) {
```

Comment:
I was about to ask why you didn't use the handler I implemented, but then realized I hadn't settled on its API and so was still in a branch on my laptop. Doh. I hadn't considered the use of a channel though, maybe that will clear it up.

Anyway, this is a fine start, can port to the more feature-complete handler later.